routing/result_interpretation.go

```diff
@@ -95,44 +100,49 @@ func interpretResult(rt *mcRoute, success bool, failureSrcIdx *int,
 func (i *interpretedResult) processSuccess(route *mcRoute) {
 	// For successes, all nodes must have acted in the right way. Therefore
 	// we mark all of them with a success result.
-	i.successPairRange(route, 0, len(route.hops)-1)
+	i.successPairRange(route, 0, len(route.hops.Val)-1)
 }
 
 // processFail processes a failed payment attempt.
-func (i *interpretedResult) processFail(rt *mcRoute, errSourceIdx *int,
-	failure lnwire.FailureMessage) {
+func (i *interpretedResult) processFail(rt *mcRoute,
+	errSourceIdx fn.Option[uint8], failure fn.Option[failureMessage]) {
+
+	var idx int
+	indexSet := fn.MapOptionZ(errSourceIdx, func(a uint8) bool {
```

Comment:
maybe one could use `ElimOption` to handle both cases in there, the downside would be that both branches are indented one level

Reply:
did you try do this and come up with a nice way? 
I gave it a try now but I dont find this very easy to read:

```
         idx := fn.ElimOption(errSourceIdx, func() int {
		i.processPaymentOutcomeUnknown(rt)

		return -1
	}, func(u uint8) int {
		return int(u)
	})
	if idx == -1 {
		return
	}
```

but maybe you had some other way in mind?

Reply:
I thought about it using it this way (with a dummy return value), not sure how nice it is :)
```go
	fn.ElimOption(
		errSourceIdx,
		func() any {
			i.processPaymentOutcomeUnknown(rt)
			return nil
		},
		func(uidx uint8) any {
			idx := int(uidx)

			failMsg := failure.UnwrapOr(failureMessage{nil}).FailureMessage

			// If the payment was to a blinded route and we received
			// an error from after the introduction point, handle
			// this error separately - there has been a protocol
			// violation from the introduction node. This penalty
			// applies regardless of the error code that is
			// returned.
			introIdx, isBlinded := introductionPointIndex(rt)
			if isBlinded && introIdx < idx {
				i.processPaymentOutcomeBadIntro(
					rt, introIdx, idx,
				)

				return nil
			}

			switch idx {
			// We are the source of the failure.
			case 0:
				i.processPaymentOutcomeSelf(rt, failMsg)

			// A failure from the final hop was received.
			case len(rt.hops.Val):
				i.processPaymentOutcomeFinal(rt, failMsg)

			// An intermediate hop failed. Interpret the outcome,
			// update reputation and try again.
			default:
				i.processPaymentOutcomeIntermediate(
					rt, idx, failMsg,
				)
			}

			return nil
		},
	)
```

Reply:
are you ok with me keeping it as is?